contracts/fatex/FateReward/MembershipWithReward.sol

```diff
@@ -0,0 +1,268 @@
+            _pid,
+            _user,
+            true
+        ) * POINTS_PER_BLOCK;
+    }
+
+    /// @dev rank with Points and set rankedNumber to each user
+    function rank(uint256 _pid) public returns (uint256) {
+        if (isFatePool[_pid]) {
+            address[] memory pool_deposited_user_list = depositedUsers[_pid];
+            uint256 deposited_user_counts = pool_deposited_user_list.length;
+            require(deposited_user_counts > 0, "rank: no_deposited_users_yet");
+            
+
+            uint256[] memory pointsList = new uint256[](deposited_user_counts);
+            for(uint i = 0; i < deposited_user_counts; i++) {
+                pointsList[i] = userPoints(_pid, pool_deposited_user_list[i]);
+            }
+
+            uint256[] memory sortedPointsList = RankedArray.sort(pointsList);
+            for (uint i = 0; i < deposited_user_counts; i++) {
+                address userAddr = depositedUsers[_pid][i];
+                MembershipInfo memory membership = userMembershipInfo[_pid][userAddr];
+
+                userMembershipInfo[_pid][userAddr] = MembershipInfo({
+                    firstDepositBlock: membership.firstDepositBlock,
+                    rankedNumber: RankedArray.getIndex(sortedPointsList, pointsList[i]),
+                    lastWithdrawBlock: membership.lastWithdrawBlock
+                });
+            }
+
+        }
+    }
+
+    function _getBlocksOfPeriod(
+        uint256 _pid,
+        address _user,
+        bool _isDepositPeriod
+    ) internal view returns (uint256 blocksOfPeriod) {
+        if (isFatePool[_pid]) {
+            uint256 currentBlockNumber = block.number;
+            uint256 endBlock = currentBlockNumber > epochEndBlock ? epochEndBlock : currentBlockNumber;
+            MembershipInfo memory membership = userMembershipInfo[_pid][_user];
+            uint256 startBlock = _isDepositPeriod ? membership.firstDepositBlock : membership.lastWithdrawBlock;
+            
+            if (endBlock >= startBlock) {
+                blocksOfPeriod = endBlock - startBlock;
+            }
+        }
+    }
+
+    /// @dev calculate lockedRewardsFees as percent that will be sent to the rewardController
+    function _getLockedRewardsFeePercent(
+        uint256 _pid,
+        address _caller
+    ) internal view returns(uint256 percent) {
+        if (isExcludedAddress[_caller]) {
+            percent = 0;
+        } else {
+            percent = lockedRewardsFeePercents[
+                RankedArray.getIndexOfBlocks(
+                    _getBlocksOfPeriod(
+                        _pid,
+                        _caller,
+                        true
+                    ),
+                    lockedRewardsPeriodBlocks
+                )
+            ];
+        }
+    }
+
+    /// @dev calculate lpWithdrawFees as percent that will be sent to the rewardController
+    function _getLPWithdrawFeePercent(uint256 _pid, address _caller) internal view returns(uint256 percent) {
```

Comment:
Add better documentation - is the `percent` returned the percent kept by the user or the percent paid from their LP balance?